### ISSUE TITLE
Enabling Specialization Augments to modify machine logic or even introduce custom machine logic

### ISSUE THREAD

**Issue by Drullkus** [CoFH/Feedback#20]
What'd I like to see are Specialization Augments being able introduce custom machine logic. Currently: You can change the stone output of the Igneous Extruder to Andesite, Diorite, or Granite. You can use the Numismatic Press augment to turn your Compactor into a minting machine. You can make the Pulverizer require Petrotheum to increase earnings, or Induction Smelter require Pyrotheum for similar purposes.

What I see being able to enhance the feature of Specialization Augments to other modders is allowing them to also be able to introduce their own Specialization Augments that will also modify behavior of machines. How this can happens is modifying the machine logic directly. You can provide a class for the logic for the machine, and allow the modder to extend it into a finalized logic object that a machine can use. This class can be abstract or not, that's up to you.

I have a couple of made-up use-case scenarios followed by an actual use-case that I am hoping to use.

Let's say I want an Induction Smelter to manufacture me concrete. I'd introduce custom machine logic through my own Specialization Augment. This makes use of the hidden-but-now-visible fluid slot, and allows it to accept water. I can also use two of any following blocks: stone, andesite, diorite, or granite. The output will be a concrete block made with some water and then a combined visual similar to the two chosen blocks. I can have an almost-colorless olive-green concrete block, or I can have a red and gray concrete block.

What's actually happening here is that the machine logic class is inferring the color of the resulting block based on the first two blocks. The logic is writing the color to the NBT of the resulting item stack when it has finished manufacturing. It will then consume one each of those two blocks as well as some water. It has ignored the Induction Smelter recipe map, that's fine.

Let's make the Numismatic Dynamo generate power each time a player crouches nearby. There's something dirty in the minds of the inhabitants who live on the other side in the Dynamo, and they enjoy the visual. They give a tiny bit of power each time the player crouches, rewarding the player for innocent acts but morsel enough to make the player want more. The amount of power given exponentially scales with amount of players rapidly pressing the crouch key.

Behind the scenes, the normal logic of the dynamo that allows it to consume minted metals is replaced, and the slot to the left of the power meter is hidden. The logic class the dynamo uses the location of the dynamo to scan for players in a small radius. It tracks each of them in the radius, and checks to see if they are sneaking but were not sneaking the tick before. If this condition is met, the dynamo makes only a bit of RF. 

What I want to implement if this suggestion is implemented is a Specialization Augment for the Magma Crucible that converts it into a Thermal Smeltery. It invalidates regular Magma Crucible operations and allows it run the Tinker's Construct recipe maps instead. It will process the items into molten metal like a Tinker's Construct Smeltery. Unlike the regular Magma Crucible, it can infer recipes based on the NBT of the ItemStack as well since it is using a NBT-sensitive recipe map.

A basic machine logic class can be the machine's tile entity class directly, or it can be an object used by the tile entity. Some machines have a hidden fluid gauge and it is up to the logic class's boolean whether it should be rendered or not. If it is not rendered, then it should never be used and blocked off from usage. The Specialization Augment item should be registered in a special registry where the augment item is paired with a reference to the custom machine logic, so the custom machine logic object can be operated when the Specialization Augment is installed into the appropriate machine.

Thank you for reading my suggestion! :)

TL;DR I'm suggesting that other modders can add Specialization Augments that introduce custom Logic Objects that Machines can use for running customized operations instead of their regular behavior.

**Comment by SnowShock35** [comment]
> What'd I like to see are Specialization Augments being able introduce custom machine logic. Currently: You can change the stone output of the Igneous Extruder to Andesite, Diorite, or Granite. You can use the Numismatic Press augment to turn your Compactor into a minting machine. You can make the Pulverizer require Petrotheum to increase earnings, or Induction Smelter require Pyrotheum for similar purposes.
> 
> What I see being able to enhance the feature of Specialization Augments to other modders is allowing them to also be able to introduce their own Specialization Augments that will also modify behavior of machines. How this can happens is modifying the machine logic directly. You can provide a class for the logic for the machine, and allow the modder to extend it into a finalized logic object that a machine can use. This class can be abstract or not, that's up to you.
> 
> I have a couple of made-up use-case scenarios followed by an actual use-case that I am hoping to use.
> 
> Let's say I want an Induction Smelter to manufacture me concrete. I'd introduce custom machine logic through my own Specialization Augment. This makes use of the hidden-but-now-visible fluid slot, and allows it to accept water. I can also use two of any following blocks: stone, andesite, diorite, or granite. The output will be a concrete block made with some water and then a combined visual similar to the two chosen blocks. I can have an almost-colorless olive-green concrete block, or I can have a red and gray concrete block.
> 
> What's actually happening here is that the machine logic class is inferring the color of the resulting block based on the first two blocks. The logic is writing the color to the NBT of the resulting item stack when it has finished manufacturing. It will then consume one each of those two blocks as well as some water. It has ignored the Induction Smelter recipe map, that's fine.
> 
> Let's make the Numismatic Dynamo generate power each time a player crouches nearby. There's something dirty in the minds of the inhabitants who live on the other side in the Dynamo, and they enjoy the visual. They give a tiny bit of power each time the player crouches, rewarding the player for innocent acts but morsel enough to make the player want more. The amount of power given exponentially scales with amount of players rapidly pressing the crouch key.
> 
> Behind the scenes, the normal logic of the dynamo that allows it to consume minted metals is replaced, and the slot to the left of the power meter is hidden. The logic class the dynamo uses the location of the dynamo to scan for players in a small radius. It tracks each of them in the radius, and checks to see if they are sneaking but were not sneaking the tick before. If this condition is met, the dynamo makes only a bit of RF.
> 
> What I want to implement if this suggestion is implemented is a Specialization Augment for the Magma Crucible that converts it into a Thermal Smeltery. It invalidates regular Magma Crucible operations and allows it run the Tinker's Construct recipe maps instead. It will process the items into molten metal like a Tinker's Construct Smeltery. Unlike the regular Magma Crucible, it can infer recipes based on the NBT of the ItemStack as well since it is using a NBT-sensitive recipe map.
> 
> A basic machine logic class can be the machine's tile entity class directly, or it can be an object used by the tile entity. Some machines have a hidden fluid gauge and it is up to the logic class's boolean whether it should be rendered or not. If it is not rendered, then it should never be used and blocked off from usage. The Specialization Augment item should be registered in a special registry where the augment item is paired with a reference to the custom machine logic, so the custom machine logic object can be operated when the Specialization Augment is installed into the appropriate machine.
> 
> Thank you for reading my suggestion! :)
> 
> TL;DR I'm suggesting that other modders can add Specialization Augments that introduce custom Logic Objects that Machines can use for running customized operations instead of their regular behavior.

### Tl;Dr Igneous Extruder upgrade that gives marble, limestone and basalt from Chisel! 😜

**Comment by SnowShock35** [comment]
In all seriousness, I really do like the idea of additional augments (That don't break balance e.g. speed modifiers or amount modifiers). It would be a great way for mods to accompany Thermal * mods and could be interesting to see what modders could integrate with this ability. I know from Drullkus it would enable him to make something like Thermal Smeltery a much cooler experience by adding an augment instead of just adding recipes. 

Tl;Dr, I :heart: the idea

**Comment by codewarrior0** [comment]
Wow... well, I would have just added Extruder augments whose NBT tags tell the Extruder which block to replace Stone with.

Delegating the crafting logic to a class associated with the augment would also accomplish that, but it seems like it would be no small task to actually go and implement it for each of the machines (and dynamos?).

**Comment by KingLemming** [comment]
You're asking a lot here.

Like, I think it will literally be less code for me to just add augments that do what you want than write a system which there's a very high likelihood will only be used a handful of times by a small number of modders.

I'm not saying no, I'm just saying that it's incredibly complicated.

**Comment by Skillotic0703** [comment]
I think he wants that to be able to melt tinkers' construct items in the magma crucible. In 1.7.10, the thermal smeltery mod added two new machines. The first one allowed to mold molten metals into ingots (or blocks) and the second one to mold molten metals into tinkers' construct parts.

**Comment by KingLemming** [comment]
Alright so I've given this a lot of thought recently, and it just isn't happening, sorry. It'd be a rewrite of tremendous work for frankly not much additional capability. And I'm not even sure it's possible, given the container/gui interactions, packets, etc.
